cli/src/user.rs

```diff
@@ -389,88 +506,94 @@ impl User {
-        };
+        {
+            // Build a proposal with this key package and do the MLS bits.
+            let mut groups = self.groups.borrow_mut();
+            let group = match groups.get_mut(&group_name) {
+                Some(g) => g,
+                None => return Err(format!("No group with name {group_name} known.")),
+            };
 
-        let (out_messages, welcome, _group_info) = group
-            .mls_group
-            .borrow_mut()
-            .add_members(
-                &self.crypto,
-                &self.identity.borrow().signer,
-                &[joiner_key_package.into()],
-            )
-            .map_err(|e| format!("Failed to add member to group - {e}"))?;
-
-        /* First, send the MlsMessage commit to the group.
-        This must be done before the member invitation is locally committed.
-        It avoids the invited member to receive the commit message (which is in the previous group epoch).*/
-        log::trace!("Sending commit");
-        let group = groups.get_mut(group_id).unwrap(); // XXX: not cool.
-        let group_recipients = self.recipients(group);
-
-        let msg = GroupMessage::new(out_messages.into(), &group_recipients);
-        self.backend.send_msg(&msg)?;
-
-        // Second, process the invitation on our end.
-        group
-            .mls_group
-            .borrow_mut()
-            .merge_pending_commit(&self.crypto)
-            .expect("error merging pending commit");
+            let (out_messages, welcome, _group_info) = group
+                .mls_group
+                .borrow_mut()
+                .add_members(
+                    &self.crypto,
+                    &self.identity.borrow().signer,
+                    &[joiner_key_package.into()],
+                )
+                .map_err(|e| format!("Failed to add member to group - {e}"))?;
+
+            /* First, send the MlsMessage commit to the group.
+            This must be done before the member invitation is locally committed.
+            It avoids the invited member to receive the commit message (which is in the previous group epoch).*/
+            log::trace!("Sending commit");
+            let group = groups.get_mut(&group_name).unwrap(); // XXX: not cool.
+            let group_recipients = self.recipients(group);
+
+            let msg = GroupMessage::new(out_messages.into(), &group_recipients);
+            self.backend.send_msg(&msg)?;
+
+            // Second, process the invitation on our end.
+            group
+                .mls_group
+                .borrow_mut()
+                .merge_pending_commit(&self.crypto)
+                .expect("error merging pending commit");
+
+            // Finally, send Welcome to the joiner.
+            log::trace!("Sending welcome");
+            self.backend
+                .send_welcome(&welcome)
+                .expect("Error sending Welcome message");
+        }
```

Comment:
I think you can just use `drop(groups)` here instead of putting everything into a block. That would certainly make the diff look better, but I otherwise don't have a strong preference. (Same below.)

Reply:
done